src/bootstrap/src/utils/bin_helpers.rs

```diff
@@ -19,10 +23,29 @@ pub(crate) fn parse_rustc_verbose() -> usize {
 ///
 /// If "RUSTC_STAGE" was not set, the program will be terminated with 101.
 pub(crate) fn parse_rustc_stage() -> String {
-    std::env::var("RUSTC_STAGE").unwrap_or_else(|_| {
+    env::var("RUSTC_STAGE").unwrap_or_else(|_| {
         // Don't panic here; it's reasonable to try and run these shims directly. Give a helpful error instead.
         eprintln!("rustc shim: FATAL: RUSTC_STAGE was not set");
         eprintln!("rustc shim: NOTE: use `x.py build -vvv` to see all environment variables set by bootstrap");
         std::process::exit(101);
     })
 }
+
+/// Writes the command invocation to a file if `DUMP_BOOTSTRAP_SHIMS` is set during bootstrap.
+///
+/// Before writing it, replaces user-specific values to create generic dumps for cross-environment
+/// comparisons.
+pub(crate) fn maybe_dump(dump_name: String, cmd: &Command) {
+    if let Ok(dump_dir) = env::var("DUMP_BOOTSTRAP_SHIMS") {
+        let dump_file = format!("{dump_dir}/{dump_name}");
+
+        let mut file =
+            OpenOptions::new().create(true).write(true).append(true).open(&dump_file).unwrap();
+
+        let cmd_dump = format!("{:?}\n", cmd);
+        let cmd_dump = cmd_dump.replace(&env::var("BUILD_OUT").unwrap(), "${BUILD_OUT}");
```

Comment:
To keep this a bit more machine-readable, I wonder if we should instead just print out the command verbatim, and then also print out the environment variables and arguments with which the shim was invoked (e.g. in JSON), and have a postprocessing step that will normalize this file and check it against a golden version?

Reply:
Wouldn't that make it more complicated? Currently each invocation along with its environment variables is recorded per line, making it human readable and easy to compare. When we separate the command and its environment variables in dumps we will need to find a way to map them when we want to match the rustc invocation with its environment variables.

Regarding the golden file, as a start, these dumps are mainly intended for manual comparisons rather than automatic handling. We need to figure out some challenges; for example, right now we can't compare across different architectures such as from x86 to arm, because the invocations will be different. Also, blessing golden files to keep them synchronized will take quite some time, as these dumps are created in a non-dry mode. I think we should not go to that path as a start until we figure out practical solutions for these challanges.

Reply:
Ah, I see. Yeah, so if it's intended for manual comparison, and replacing the env. vars is the only operation that will be currently performed on these dumps, then let's just keep it simple for now.